content/en/docs/File Format/implementationstatus.md

```diff
@@ -115,12 +115,12 @@ Implementations:
 
 | Format                                       | C++   | Java  | Go    | Rust  | cuDF  |
 | -------------------------------------------- | ----- | ----- | ----- | ----- | ----- |
-| External column data (1)                     |  ✅   |  ✅   |       |       |  (W)  |
-| Row group "Sorting column" metadata (2)      |  ✅   |  ❌   |       |       |  (W)  |
-| Row group pruning using statistics           |  ❌   |  ✅   |       |       |  ✅   |
-| Row group pruning using bloom filter         |  ❌   |  ✅   |       |       |  ✅   |
-| Reading select columns only                  |  ✅   |  ✅   |       |       |  ✅   |
-| Page pruning using statistics                |  ❌   |  ✅   |       |       |  ❌   |
+| External column data (1)                     |  ✅   |  ✅   |       |  ❌   |  (W)  |
+| Row group "Sorting column" metadata (2)      |  ✅   |  ❌   |       |  ✅   |  (W)  |
+| Row group pruning using statistics           |  ❌   |  ✅   |       |  ✅   |  ✅   |
+| Row group pruning using bloom filter         |  ❌   |  ✅   |       |  ✅   |  ✅   |
+| Reading select columns only                  |  ✅   |  ✅   |       |  ✅   |  ✅   |
+| Page pruning using statistics                |  ❌   |  ✅   |       |  ✅   |  ❌   |
```

Comment:
I'm not entirely sure how to do the pruning ones...IIUC parquet-rs allows for pruning, but the actual work needs to be done outside the library. Perhaps these should be Xs like C++?

Reply:
I think it is fair to say pruning is supported, in that the APIs to do it are there, it just isn't batteries included (we don't ship an expression engine)

Reply:
I agree we should mark parquet-rs as supporting pruning

Speicficially this structure gets the statistics as arrow record batches (either pages or row groups)
- https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/statistics/struct.StatisticsConverter.html

And then you can specify which row groups to read read via
- https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/struct.ArrowReaderBuilder.html#method.with_row_groups
- https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/struct.ArrowReaderBuilder.html#method.with_row_selection

As @tustvold  says parquet-rs doesn't provide a way to evaluate an expression on those arrow arrays, but  you can use a query engine (like DataFusion!) to do so

Reply:
BTW I wonder if we should propose adding a row for "predicate pushdown" (aka evaluating predicates based on scans) -- basically what https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/trait.ArrowPredicate.html provides

Reply:
I believe "late materialization" is the precise technique, and I agree it would be good to add a row for this